### ISSUE TITLE
Missing depth24unorm fomat?

### ISSUE THREAD

**Issue by Kangz** [gpuweb/gpuweb#2732]
The [depth-specific format](https://gpuweb.github.io/gpuweb/#aspect-specific-format) for `depth24unorm-stencil8` is currently `depth24plus`. This requires that when the feature with `depth24unorm-stencil8` is exposed, that `depth24plus` be exactly `depth24unorm`.

It seems that instead we're just missing a format that's `depth24unorm` and only available when the feature is enabled.

**Comment by Kangz** [comment]
And we might want to specify that `depth24unorm(-stencil8)` T<->B copies are allowed since the format is known. The only question is when copying `depth24unorm` to a buffer, what is the value of the missing stencil bytes. Vulkan says it is undefined.

**Comment by Kangz** [comment]
Tentatively setting v1 since the optional feature is already in the spec.

**Comment by haoxli** [comment]
[D3D12](https://microsoft.github.io/DirectX-Specs/d3d/PlanarDepthStencilDDISpec.html#dxgi_format_r24g8_typeless) says the stencil bytes must be initialized to **zero** when copying from the depth plane of a `DXGI_FORMAT_R24G8_TYPELESS` texture. [Metal](https://developer.apple.com/documentation/metal/mtlpixelformat/mtlpixelformatdepth24unorm_stencil8)  sets the stencil bytes to arbitrary data, like Vulkan did. 

I did not find any formats in Metal can be used for `depth24unorm` format (D3D12 has `DXGI_FORMAT_R24_UNORM_X8_TYPELESS`, Vulkan has `VK_FORMAT_X8_D24_UNORM_PACK32`), which may block us.

**Comment by kainino0x** [comment]
I'm guessing - without having thought about it too deeply - that we could use Depth24Unorm_Stencil8 under the hood.

However that doesn't fix the undefined behavior for T<->B copies. I don't see a way to avoid that except to emulate using a compute shader, which the user can do themselves as described in the non-normative note here: https://gpuweb.github.io/gpuweb/#copying-depth-stencil

**Comment by Kangz** [comment]
I'm reasonably confident that all the special cases surrounding depth24unorm-stencil8 in the various APIs come from the fact that this format may be interleaved or multiplanar depending on the hardware:

 - If the format is interleaved, data from the stencil buffer gets written to the buffer in T->B copies, which means that the padding byte should be filled with the stencil for that pixel.
 - If the format is truly multiplanar:
   - if the depth plane is aligned to 4, there's an extra padding byte that gets copied to and from in T<->B copies.
   - if the depth plane is packed (it is workable because in most cases you load quads of pixels, so that's 12 bytes which are aligned to 4), then the padding byte is discarded in B->T copies, and not most likely written to 0 in T->B (it could be not written but that would require 3byte memory writes, or 2 + 1, which would be odd / slow).

So while the various specs say that the padding byte may be arbitrary, I don't think it is in practice: it will be either 0, or whatever data you copied into it with a previous B->T (or T (depth24unorm-stencil8)->T (Depth24unorm)) operation. IMHO it seems fine to spec that it is arbitrary with a note explaining the above, this is kind of a corner case. The only thing implementations have to be careful about is to zero-initialize the storage when the HW uses an interleaved format. This can be done either by initializing with a copy, or by using depth24unorm-stencil8 under the hood.

So I'd like to resolve to say that the padding byte in B->T is ignored and arbitrary in T->B with a note explaining the above.

**Comment by Kangz** [comment]
Worst case this is an optional feature so implementation that are not comfortable doing this on a specific piece of hardware can just not expose the feature.

**Comment by kainino0x** [comment]
> So while the various specs say that the padding byte may be arbitrary, I don't think it is in practice: it will be either 0, or whatever data you copied into it with a previous B->T (or T (depth24unorm-stencil8)->T (Depth24unorm)) operation.

This may be true, but I'm not that confident about it being universal. For example I can imagine you have some region of a cache used as an intermediate for a copy, and the device does the copy in two steps: one from T to cache initializing ONLY the first 24 and leaving garbage for the last 8, and then one from cache to the buffer copying the garbage 8 bits.

**Comment by Kangz** [comment]
Right, we can never make too many assumptions about GPU architectures. Maybe implementations can verify things through the WebGPU CTS tests and if need be run a compute shader just to clear the padding bytes after the copy?

**Comment by kainino0x** [comment]
I think we would want pretty robust testing if we were to encode this assumption in the spec/tests, and the implementation workarounds wouldn't be trivial. Maybe enabling this copy makes more sense for polish-post-v1?

**Comment by Kangz** [comment]
Happy to have this move to polish-post-v1 but because it is a likely bugfix for the `depth24unorm-stencil8` feature, I think we'll need to move that whole feature to polish-post-v1.

**Comment by kainino0x** [comment]
In today's discussion I think the main question we hit is whether `depth24unorm-stencil8` actually has any value if it's not copyable. We added it in #1024. That's linked to #755, but that issue was actually only about `depth32float-stencil8`. I think we added `depth24unorm-stencil8` basically to have a concrete format for it. But anytime it's available, implementations are most likely going to use it to implement `depth24plus-stencil8` anyway because it takes less memory (4 bytes instead of 5-8 bytes). If it doesn't add any capabilities over `depth24plus-stencil8`, the only real difference is that it guarantees a specific precision instead of being undefined. We can solve that just as well with #1887 which would tell you how `depth24plus-stencil8` is implemented.

Given that, I think that it does make most sense to remove the feature for v1.

**Comment by kainino0x** [comment]
Opened https://github.com/gpuweb/gpuweb/pull/2950 in case we decide we want to do that.

**Comment by Kangz** [comment]
Agreed that `depth24unorm-stencil8` isn't really useful if it doesn't have more capabilities than `depth24plus-stencil8`. However it seems useful to have depth stencil formats that you can copy from/to buffers and it's not possible without a concrete format (or compute shaders / render passes to do the copy).

So I'm happy to remove it for v1, but we should consider it for post-v1 (that or allowing copies between buffers and depth aspects in other ways)

**Comment by Kangz** [comment]
Fixed by #2950
